ui/src/routes/devices.$id.settings.macros.tsx

```diff
@@ -0,0 +1,1439 @@
+                <div className="flex flex-wrap gap-1">
+                  {mods.map(option => (
+                    <label 
+                      key={option.value} 
+                      className={`flex items-center px-2 py-1 rounded border cursor-pointer text-xs font-medium transition-colors ${
+                        ensureArray(step.modifiers).includes(option.value) 
+                          ? 'bg-blue-100 border-blue-300 text-blue-700 dark:bg-blue-900/40 dark:border-blue-600 dark:text-blue-200' 
+                          : 'bg-slate-100 border-slate-200 text-slate-600 hover:bg-slate-200 dark:bg-slate-800 dark:border-slate-700 dark:text-slate-300 dark:hover:bg-slate-700'
+                      }`}
+                    >
+                      <input
+                        type="checkbox"
+                        className="sr-only"
+                        checked={ensureArray(step.modifiers).includes(option.value)}
+                        onChange={e => {
+                          const modifiersArray = ensureArray(step.modifiers);
+                          const newModifiers = e.target.checked
+                            ? [...modifiersArray, option.value]
+                            : modifiersArray.filter(m => m !== option.value);
+                          onModifierChange(newModifiers);
+                        }}
+                      />
+                      {option.label.split(' ')[1] || option.label}
+                    </label>
+                  ))}
+                </div>
+              </div>
+            ))}
+          </div>
+        </div>
+        
+        <div className="w-full flex flex-col gap-1">
+          <label className="text-sm font-medium text-slate-700 dark:text-slate-300">
+            Keys:
+          </label>
+          
+          <div className="macro-key-group flex flex-wrap gap-1 mb-2">
+            {ensureArray(step.keys).map((key, keyIndex) => (
+              <span
+                key={keyIndex}
+                className="macro-key-badge inline-flex items-center rounded-md bg-blue-100 px-2 py-1 text-xs font-medium text-blue-700 dark:bg-blue-900/40 dark:text-blue-200"
+              >
+                {key}
+                <button
+                  type="button"
+                  className="ml-1 text-xs text-blue-500 hover:text-blue-700 dark:text-blue-400 dark:hover:text-blue-300"
+                  onClick={() => {
+                    const newKeys = ensureArray(step.keys).filter((_, i) => i !== keyIndex);
+                    onKeySelect({ value: null, keys: newKeys });
+                  }}
+                >
+                  ×
+                </button>
+              </span>
+            ))}
+          </div>
+
+          <KeyCombobox
+            stepIndex={stepIndex}
+            step={step}
+            onSelect={onKeySelect}
+            query={keyQuery}
+            onQueryChange={onKeyQueryChange}
+            getFilteredOptions={getFilteredKeys}
```

Comment:
It took me some time to understand the logic behind this dropdown when using the UI for the first time. 

Add a map for the label in the Combobox. `KeyA` should be `A`, `NumPad0` should be `Num Pad 0` etc...

Reply:
pulled out some of the logic from the VirtualKeyboard component and added it back down into keyboardMappings.ts file. lmk if you want a different approach